op-conductor/rpc/ws/server.go

```diff
@@ -0,0 +1,234 @@
+	}
+}
+
+// run starts the hub's main loop
+func (h *Hub) run() {
+	for {
+		select {
+		case <-h.done:
+			// Close all remaining client connections
+			for client := range h.clients {
+				client.Close()
+				delete(h.clients, client)
+			}
+			return
+		case client := <-h.register:
+			h.clients[client] = true
+			clientCount := len(h.clients)
+			h.log.Info("Client registered with hub", "totalClients", clientCount)
+		case client := <-h.unregister:
+			if _, ok := h.clients[client]; ok {
+				delete(h.clients, client)
+				client.Close()
+			}
+		case message := <-h.broadcast:
+			successCount := 0
+			dropCount := 0
+
+			for client := range h.clients {
+				select {
+				case client.send <- message:
+					// Message sent successfully
+					successCount++
+				default:
+					// Channel is full, client is likely slow/dead
+					// The ping mechanism will detect and clean up dead clients
+					h.log.Debug("Failed to send message to client, channel full")
+					dropCount++
+				}
+			}
+			if dropCount > 0 {
+				h.log.Warn("Failed to send message to all clients, dropped", "successCount", successCount, "dropCount", dropCount)
+			}
+		}
+	}
+}
+
+// Client represents a connected WebSocket client
+type Client struct {
+	conn   *websocket.Conn
+	send   chan []byte
+	ctx    context.Context
+	cancel context.CancelFunc
+	hub    *Hub
+	log    log.Logger
+	mu     sync.Mutex
+}
+
+// Close closes the client's WebSocket connection and send channel
+func (c *Client) Close() {
+	// this mutex is used to prevent concurrent close operations
+	// double close will panic
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.cancel()
+	c.conn.Close(websocket.StatusNormalClosure, "client closed")
+	close(c.send)
+}
+
+// newClient creates a new client with default settings
+func newClient(conn *websocket.Conn, ctx context.Context, hub *Hub, logger log.Logger) *Client {
+	ctx, cancel := context.WithCancel(ctx)
+	return &Client{
+		conn:   conn,
+		send:   make(chan []byte, sendChannelBufferSize),
+		ctx:    ctx,
+		cancel: cancel,
+		hub:    hub,
+		log:    logger,
+	}
+}
+
+// serveWs handles WebSocket requests from clients
+func (h *Handler) serveWs(w http.ResponseWriter, r *http.Request) {
+	// Upgrade the HTTP connection to a WebSocket connection using coder/websocket
+	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
+		CompressionMode: websocket.CompressionDisabled,
+	})
+
+	if err != nil {
+		h.log.Error("failed to upgrade connection", "err", err)
+		return
+	}
+
+	client := newClient(conn, r.Context(), h.hub, h.log)
+
+	// Register the client with the hub
+	h.hub.register <- client
+	h.log.Info("WebSocket client connected")
+
+	// Start client handling
+	go h.writePump(client)
+	h.readPump(client)
+}
+
+// readPump for followers where you don't expect regular data messages
+func (h *Handler) readPump(client *Client) {
+	defer func() {
+		// Unregister the client when the read pump exits
+		h.hub.unregister <- client
```

Comment:
unregister is called twice for read/write Pump, I'm thinking maybe just do it in readPump